extensions/amp-document-recommendations/0.1/amp-document-recommendations.js

```diff
@@ -220,15 +230,42 @@ export class AmpDocumentRecommendations extends AMP.BaseElement {
       this.appendDivision_();
       this.appendArticleLinks_(this.nextArticle_ + 1);
     });
+
+    this.viewport_.onScroll(() => this.scrollHandler_());
+    this.viewport_.onResize(() => this.scrollHandler_());
   }
 
   /** @override */
   layoutCallback() {
-    // TODO(emarchiori): Decide when to append the next article
-    // based on PositionObserver.
-    this.appendNextArticle_();
     return Promise.resolve();
   }
+
+  /**
+   * Handles scroll events from the viewport and appends the next document when
+   * the user comes within {@code PRERENDER_VIEWPORT_COUNT} viewports of the
+   * end.
+   * @private
+   */
+  scrollHandler_() {
+    if (this.documentQueued_) {
+      return;
+    }
+
+    const viewportSize = this.viewport_.getSize();
+    const viewportBox =
+        layoutRectLtwh(0, 0, viewportSize.width, viewportSize.height);
+    this.viewport_.getClientRectAsync(this.element).then(elementBox => {
+      if (this.documentQueued_) {
```

Comment:
In what situation would the first `this.documentQueued_` check be `true`, but the second be `false`?

Reply:
It's async, so I can imagine multiple calls to getClientRectAsync() being queued up before the first has resolved, or would that not be possible?

Reply:
Makes sense. I'm not sure how I'd refactor to simplify.